File: internal/modules/periode/domain/periode.go
package domain

import (
	"context"
	"time"
)

type PeriodeAkademik struct {
	ID        string    `json:"id" gorm:"primaryKey;type:varchar(255)"`
	Tahun     string    `json:"tahun" gorm:"type:varchar(20);not null"` // e.g. "2024/2025"
	Jenis     string    `json:"jenis" gorm:"type:varchar(10);not null"` // "ganjil", "genap", "pendek"
	IsActive  bool      `json:"is_active" gorm:"not null;default:false"`
	CreatedAt time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt time.Time `json:"updated_at" gorm:"autoUpdateTime"`
}

type CreatePeriodeRequest struct {
	Tahun string `json:"tahun" validate:"required"`
	Jenis string `json:"jenis" validate:"required,oneof=ganjil genap pendek"`
}

type PeriodeRepository interface {
	Create(ctx context.Context, p *PeriodeAkademik) error
	GetAll(ctx context.Context) ([]*PeriodeAkademik, error)
	GetByID(ctx context.Context, id string) (*PeriodeAkademik, error)
	GetActive(ctx context.Context) (*PeriodeAkademik, error)
	Delete(ctx context.Context, id string) error
	SetActive(ctx context.Context, id string) error
	DeactivateAll(ctx context.Context) error
}

type PeriodeService interface {
	Create(ctx context.Context, req CreatePeriodeRequest) (*PeriodeAkademik, error)
	GetAll(ctx context.Context) ([]*PeriodeAkademik, error)
	GetActive(ctx context.Context) (*PeriodeAkademik, error)
	Delete(ctx context.Context, id string) error
	SetActive(ctx context.Context, id string) error
}

File: frontend/src/lib/services/periode.ts
import { fetchApi } from './api';
import type { ApiResponse } from '../types';

export interface PeriodeAkademik {
	id: string;
	tahun: string;
	jenis: 'ganjil' | 'genap' | 'pendek';
	is_active: boolean;
	created_at: string;
	updated_at: string;
}

export const periodeService = {
	getAll: async () => {
		return fetchApi<ApiResponse<PeriodeAkademik[]>>('/periode');
	},
	getActive: async () => {
		return fetchApi<ApiResponse<PeriodeAkademik>>('/periode/active');
	},
	create: async (data: { tahun: string; jenis: string }) => {
		return fetchApi<ApiResponse<PeriodeAkademik>>('/periode', {
			method: 'POST',
			body: JSON.stringify(data)
		});
	},
	delete: async (id: string) => {
		return fetchApi<ApiResponse<null>>(`/periode/${id}`, {
			method: 'DELETE'
		});
	},
	activate: async (id: string) => {
		return fetchApi<ApiResponse<null>>(`/periode/${id}/active`, {
			method: 'POST'
		});
	}
};
